File: src/math.rs
use glam::{IVec4, Mat4, Vec2, Vec3};
use sdl3::rect::Point;

pub fn angles_from_vector(forward: Vec3) -> Vec2 {
    let mut yaw;
    let mut pitch;

    // forward vector points up or down
    if forward.x == 0.0 && forward.y == 0.0 {
        yaw = 0.0;
        pitch = if forward.z > 0.0 { 270.0 } else { 90.0 };
    } else {
        yaw = forward.y.atan2(forward.x).to_degrees();
        if yaw < 0.0 {
            yaw += 360.0;
        }

        pitch = (-forward.z)
            .atan2(Vec2::new(forward.x, forward.y).length())
            .to_degrees();
        if pitch < 0.0 {
            pitch += 360.0;
        }
    }

    Vec2::new(pitch, yaw)
}

pub fn angles_to_fov(view_angles: Vec2, aim_angles: Vec2) -> f32 {
    let mut delta = view_angles - aim_angles;

    if delta.x > 180.0 {
        delta.x = 360.0 - delta.x;
    }
    delta.x = delta.x.abs();

    // clamp?
    delta.y = ((delta.y + 180.0) % 360.0 - 180.0).abs();

    delta.length()
}

pub fn vec2_clamp(vec: &mut Vec2) {
    if vec.x > 89.0 && vec.x <= 180.0 {
        vec.x = 89.0;
    }
    if vec.x > 180.0 {
        vec.x -= 360.0;
    }
    if vec.x < -89.0 {
        vec.x = -89.0;
    }
    vec.y = (vec.y + 180.0) % 360.0 - 180.0;
}

pub fn world_to_screen(screen_size: IVec4, view_matrix: Mat4, position: Vec3) -> Option<Point> {
    let mut screen_position = Vec2::new(
        view_matrix.x_axis.x * position.x
            + view_matrix.x_axis.y * position.y
            + view_matrix.x_axis.z * position.z
            + view_matrix.x_axis.w,
        view_matrix.y_axis.x * position.x
            + view_matrix.y_axis.y * position.y
            + view_matrix.y_axis.z * position.z
            + view_matrix.y_axis.w,
    );

    let w = view_matrix.w_axis.x * position.x
        + view_matrix.w_axis.y * position.y
        + view_matrix.w_axis.z * position.z
        + view_matrix.w_axis.w;

    if w < 0.01 {
        return None;
    }

    screen_position.x /= w;
    screen_position.y /= w;

    let x = screen_size.z as f32 / 2.0;
    let y = screen_size.w as f32 / 2.0;

    screen_position.x = x + 0.5 * screen_position.x * screen_size.z as f32 + 0.5;
    screen_position.y = y - 0.5 * screen_position.y * screen_size.w as f32 + 0.5;

    Some(Point::new(
        screen_position.x as i32,
        screen_position.y as i32,
    ))
}

File: src/math.cpp
#include "math.hpp"

#include <mithril/numbers.hpp>
#include "globals.hpp"

f32 ToDegrees(const f32 value) {
    return value * 180.0f / numbers::pi<f32>();
}

glm::vec2 AnglesFromVector(const glm::vec3& forward) {
    f32 yaw = 0.0f;
    f32 pitch = 0.0f;

    if (forward.x == 0.0f && forward.y == 0.0f) {
        // Straight up or down
        yaw = 0.0f;
        pitch = (forward.z > 0.0f) ? 270.0f : 90.0f;
    } else {
        yaw = ToDegrees(std::atan2(forward.y, forward.x));
        if (yaw < 0.0f) yaw += 360.0f;

        const f32 horizontal_length = glm::length(glm::vec2(forward.x, forward.y));
        pitch = ToDegrees(std::atan2(-forward.z, horizontal_length));
        if (pitch < 0.0f) pitch += 360.0f;
    }

    return {pitch, yaw};
}

f32 AnglesToFov(const glm::vec2& view_angles, const glm::vec2& aim_angles) {
    glm::vec2 delta = view_angles - aim_angles;

    delta.x = std::fabs((delta.x > 180.0f) ? (360.0f - delta.x) : delta.x);
    delta.y = std::fabs(std::fmod(delta.y + 180.0f, 360.0f) - 180.0f);

    return glm::length(delta);
}

void Vec2Clamp(glm::vec2& vec) {
    if (vec.x > 89.0f && vec.x <= 180.0f) {
        vec.x = 89.0f;
    } else if (vec.x > 180.0f) {
        vec.x -= 360.0f;
    }

    if (vec.x < -89.0f) {
        vec.x = -89.0f;
    }

    vec.y = std::fmod(vec.y + 180.0f, 360.0f) - 180.0f;
}

std::optional<glm::vec2> WorldToScreen(const glm::vec3& position) {
    glm::vec4 clip_space {
        view_matrix[0].x * position.x + view_matrix[0].y * position.y + view_matrix[0].z * position.z + view_matrix[0].w,
        view_matrix[1].x * position.x + view_matrix[1].y * position.y + view_matrix[1].z * position.z + view_matrix[1].w,
        0.0f, // not used
        view_matrix[3].x * position.x + view_matrix[3].y * position.y + view_matrix[3].z * position.z + view_matrix[3].w
    };

    if (clip_space.w < 0.01f) {
        return std::nullopt;
    }

    glm::vec2 ndc {
        clip_space.x / clip_space.w,
        clip_space.y / clip_space.w
    };

    const glm::vec2 half_screen {window_size.z * 0.5f, window_size.w * 0.5f};

    glm::vec2 screen_pos {
        half_screen.x + ndc.x * half_screen.x + 0.5f + window_size.x,
        half_screen.y - ndc.y * half_screen.y + 0.5f + window_size.y
    };

    const bool in_bounds =
        screen_pos.x >= window_size.x &&
        screen_pos.x <= window_size.x + window_size.z &&
        screen_pos.y >= window_size.y &&
        screen_pos.y <= window_size.y + window_size.w;

    if (!in_bounds) {
        return std::nullopt;
    }

    return screen_pos;
}
